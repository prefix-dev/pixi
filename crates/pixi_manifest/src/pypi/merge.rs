use indexmap::IndexMap;
use std::hash::Hash;

/// A trait for types that support a union-style merge policy.
pub trait MergeUnion {
    fn union(&self, other: &Self) -> Self;
}

/// Merge two optional single-assignment values.
/// - If both are `Some`, return a conflict error via the provided closure.
/// - Otherwise prefer `a` when set, else `b`.
pub fn merge_single_option<T: Clone, E>(
    a: &Option<T>,
    b: &Option<T>,
    conflict: impl FnOnce(&T, &T) -> E,
) -> Result<Option<T>, E> {
    match (a, b) {
        (Some(a), Some(b)) => Err(conflict(a, b)),
        (Some(a), None) => Ok(Some(a.clone())),
        (None, Some(b)) => Ok(Some(b.clone())),
        (None, None) => Ok(None),
    }
}

/// Merge two optional lists by concatenating and de-duplicating, preserving order.
pub fn merge_list_dedup<T: Clone + Eq + Hash>(
    a: &Option<Vec<T>>,
    b: &Option<Vec<T>>,
) -> Option<Vec<T>> {
    match (a, b) {
        (Some(a), Some(b)) => {
            let set = a
                .iter()
                .cloned()
                .chain(b.iter().cloned())
                .collect::<indexmap::IndexSet<_>>();
            Some(set.into_iter().collect())
        }
        (Some(a), None) => Some(a.clone()),
        (None, Some(b)) => Some(b.clone()),
        (None, None) => None,
    }
}

/// Merge two optional maps where left (`a`) overrides right (`b`) on key conflicts.
pub fn merge_map_override_left<K, V>(
    a: &Option<IndexMap<K, V>>,
    b: &Option<IndexMap<K, V>>,
) -> Option<IndexMap<K, V>>
where
    K: Clone + Eq + Hash,
    V: Clone,
{
    match (a, b) {
        (Some(a), Some(b)) => {
            let mut merged = b.clone();
            merged.extend(a.iter().map(|(k, v)| (k.clone(), v.clone())));
            Some(merged)
        }
        (Some(a), None) => Some(a.clone()),
        (None, Some(b)) => Some(b.clone()),
        (None, None) => None,
    }
}

/// Implement `MergeUnion` for `Option<T>` where `T` itself supports `MergeUnion`.
impl<T: MergeUnion + Clone> MergeUnion for Option<T> {
    fn union(&self, other: &Self) -> Self {
        // This contains the different matches for different cases
        match (self, other) {
            (Some(a), Some(b)) => Some(a.union(b)),
            (Some(a), None) => Some(a.clone()),
            (None, Some(b)) => Some(b.clone()),
            (None, None) => None,
        }
    }
}
