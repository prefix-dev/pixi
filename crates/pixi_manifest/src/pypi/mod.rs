pub mod pypi_options;
pub mod pypi_requirement;
pub mod pypi_requirement_types;
pub use pypi_requirement_types::{GitRev, PyPiPackageName, VersionOrStar};
