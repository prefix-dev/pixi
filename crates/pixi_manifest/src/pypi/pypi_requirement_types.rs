use std::{
    borrow::Borrow,
    fmt::{Display, Formatter, Write},
    str::FromStr,
};

use pep440_rs::VersionSpecifiers;
use pep508_rs::{InvalidNameError, PackageName};
use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
/// A package name for PyPI that also stores the source version of the name.
pub struct PyPiPackageName {
    source: String,
    normalized: PackageName,
}

impl Borrow<PackageName> for PyPiPackageName {
    fn borrow(&self) -> &PackageName {
        &self.normalized
    }
}

impl<'de> Deserialize<'de> for PyPiPackageName {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        serde_untagged::UntaggedEnumVisitor::new()
            .string(|str| PyPiPackageName::from_str(str).map_err(serde::de::Error::custom))
            .expecting("a string")
            .deserialize(deserializer)
    }
}

impl FromStr for PyPiPackageName {
    type Err = InvalidNameError;

    fn from_str(name: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            source: name.to_string(),
            normalized: PackageName::from_str(name)?,
        })
    }
}

impl PyPiPackageName {
    pub fn from_normalized(normalized: PackageName) -> Self {
        Self {
            source: normalized.to_string(),
            normalized,
        }
    }

    pub fn as_normalized(&self) -> &PackageName {
        &self.normalized
    }

    pub fn as_source(&self) -> &str {
        &self.source
    }
}

/// The pep crate does not support "*" as a version specifier, so we need to
/// handle it ourselves.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum VersionOrStar {
    Version(VersionSpecifiers),
    Star,
}

impl FromStr for VersionOrStar {
    type Err = pep440_rs::VersionSpecifiersParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s == "*" {
            Ok(VersionOrStar::Star)
        } else {
            Ok(VersionOrStar::Version(VersionSpecifiers::from_str(s)?))
        }
    }
}

impl Display for VersionOrStar {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            VersionOrStar::Version(v) => f.write_str(&format!("{}", v)),
            VersionOrStar::Star => f.write_char('*'),
        }
    }
}

impl From<VersionOrStar> for Option<pep508_rs::VersionOrUrl> {
    fn from(val: VersionOrStar) -> Self {
        match val {
            VersionOrStar::Version(v) => Some(pep508_rs::VersionOrUrl::VersionSpecifier(v)),
            VersionOrStar::Star => None,
        }
    }
}

impl From<VersionOrStar> for VersionSpecifiers {
    fn from(value: VersionOrStar) -> Self {
        match value {
            VersionOrStar::Version(v) => v,
            VersionOrStar::Star => VersionSpecifiers::from_iter(vec![]),
        }
    }
}

impl Serialize for VersionOrStar {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

impl<'de> Deserialize<'de> for VersionOrStar {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        VersionOrStar::from_str(&s).map_err(serde::de::Error::custom)
    }
}

#[derive(Debug, Serialize, Clone, PartialEq, Eq, Hash)]
#[serde(untagged, rename_all = "snake_case", deny_unknown_fields)]
pub enum GitRev {
    Short(String),
    Full(String),
}

impl GitRev {
    pub fn as_full(&self) -> Option<&str> {
        match self {
            GitRev::Full(full) => Some(full.as_str()),
            GitRev::Short(_) => None,
        }
    }
}

#[derive(thiserror::Error, Clone, Debug, Eq, PartialEq)]
pub enum GitRevParseError {
    #[error("Invalid length must be less than 40, actual size: {0}")]
    InvalidLength(usize),
    #[error(
        "Found invalid characters for git revision '{0}', branches and tags are not supported yet"
    )]
    InvalidCharacters(String),
}

impl FromStr for GitRev {
    type Err = GitRevParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if !s.chars().all(|c| c.is_ascii_hexdigit()) {
            return Err(GitRevParseError::InvalidCharacters(s.to_string()));
        }

        // Parse the git revision
        match s.len() {
            0..=39 => Ok(GitRev::Short(s.to_string())),
            40 => Ok(GitRev::Full(s.to_string())),
            _ => Err(GitRevParseError::InvalidLength(s.len())),
        }
    }
}

impl Display for GitRev {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            GitRev::Short(s) => f.write_str(s),
            GitRev::Full(s) => f.write_str(s),
        }
    }
}

impl<'de> Deserialize<'de> for GitRev {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s: String = Deserialize::deserialize(deserializer)?;
        if s.len() == 40 {
            Ok(GitRev::Full(s))
        } else {
            Ok(GitRev::Short(s))
        }
    }
}
#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn test_git_rev_from_str_valid_short() {
        let rev = GitRev::from_str("abc123").unwrap();
        assert_eq!(rev, GitRev::Short("abc123".to_string()));
    }

    #[test]
    fn test_git_rev_from_str_valid_full() {
        let rev = GitRev::from_str("0123456789abcdef0123456789abcdef01234567").unwrap();
        assert_eq!(
            rev,
            GitRev::Full("0123456789abcdef0123456789abcdef01234567".to_string())
        );
    }

    #[test]
    fn test_git_rev_from_str_invalid_characters() {
        let rev = GitRev::from_str("\x1b");
        assert!(rev.is_err());
        assert_eq!(
            rev.unwrap_err(),
            GitRevParseError::InvalidCharacters("\x1b".to_string())
        );
    }

    #[test]
    fn test_git_rev_from_str_invalid_length() {
        let rev = GitRev::from_str("0123456789abcdef0123456789abcdef0123456789");
        assert!(rev.is_err());
        assert_eq!(rev.unwrap_err(), GitRevParseError::InvalidLength(42));
    }
}
