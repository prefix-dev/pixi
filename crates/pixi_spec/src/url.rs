use rattler_conda_types::{package::ArchiveIdentifier, NamelessMatchSpec};
use rattler_digest::{Md5Hash, Sha256Hash};
use serde_with::serde_as;
use url::Url;

/// A specification of a package from a URL. This is used to represent both
/// source and binary packages.
#[serde_as]
#[derive(Debug, Clone, Hash, Eq, PartialEq, ::serde::Serialize, ::serde::Deserialize)]
pub struct UrlSpec {
    /// The URL of the package
    pub url: Url,

    /// The md5 hash of the package
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Md5>>")]
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the package
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Sha256>>")]
    pub sha256: Option<Sha256Hash>,
}

impl UrlSpec {
    /// Converts this instance into a [`NamelessMatchSpec`] if the URL points to
    /// a binary package.
    #[allow(clippy::result_large_err)]
    pub fn try_into_nameless_match_spec(self) -> Result<NamelessMatchSpec, Self> {
        if self.is_binary() {
            Ok(NamelessMatchSpec {
                url: Some(self.url),
                md5: self.md5,
                sha256: self.sha256,
                ..NamelessMatchSpec::default()
            })
        } else {
            Err(self)
        }
    }

    /// Converts this instance into a [`UrlSourceSpec`] if the URL points to a
    /// source package. Otherwise, returns this instance unmodified.
    #[allow(clippy::result_large_err)]
    pub fn try_into_source_url(self) -> Result<UrlSourceSpec, Self> {
        if self.is_binary() {
            Err(self)
        } else {
            Ok(UrlSourceSpec {
                url: self.url,
                md5: self.md5,
                sha256: self.sha256,
            })
        }
    }

    /// Returns true if the URL points to a binary package.
    pub fn is_binary(&self) -> bool {
        ArchiveIdentifier::try_from_url(&self.url).is_some()
    }
}

/// A specification of a source archive from a URL.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct UrlSourceSpec {
    /// The URL of the package
    pub url: Url,

    /// The md5 hash of the archive
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the archive
    pub sha256: Option<Sha256Hash>,
}

impl From<UrlSourceSpec> for UrlSpec {
    fn from(value: UrlSourceSpec) -> Self {
        Self {
            url: value.url,
            md5: value.md5,
            sha256: value.sha256,
        }
    }
}
