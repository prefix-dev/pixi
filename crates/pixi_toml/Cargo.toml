[package]
authors.workspace = true
description = "A crate that provides utilities to parse TOML files with extended diagnostics"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "pixi_toml"
readme.workspace = true
repository.workspace = true
version = "0.1.0"

[dependencies]
digest = { workspace = true }
hex = { workspace = true }
indexmap = { workspace = true }
itertools = { workspace = true }
miette = { workspace = true }
strsim = { workspace = true }
strum = { workspace = true }
toml-span = { workspace = true }

[dev-dependencies]
insta = { workspace = true }
