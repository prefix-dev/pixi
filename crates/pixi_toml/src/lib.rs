//! This crate provides utilities to work with [`toml_span`].
//!
//! Deserialization helper structs are provide to deserialize types that are not
//! directly supported by [`toml_span`].
//!
//! This crate also provides [`TomlWith`] which enables directly parsing from
//! one type into another. This makes it possible to efficiently implement
//! deserializers for types from external crates.

mod diagnostic;
mod digest;
mod from_str;
mod hash_map;
mod index_map;
mod index_set;
mod one_or_many;
mod variant;
mod with;

use std::str::FromStr;

pub use diagnostic::TomlDiagnostic;
pub use digest::TomlDigest;
pub use from_str::TomlFromStr;
pub use hash_map::TomlHashMap;
pub use index_map::TomlIndexMap;
pub use index_set::TomlIndexSet;
pub use one_or_many::OneOrMany;
use toml_span::{DeserError, Spanned, Value, de_helpers::expected, value::ValueInner};
pub use variant::TomlEnum;
pub use with::TomlWith;

/// A trait that enables efficient deserialization of one type into another.
pub trait DeserializeAs<'de, T> {
    fn deserialize_as(value: &mut Value<'de>) -> Result<T, DeserError>;
}

pub struct Same;

impl<'de, T: toml_span::Deserialize<'de>> DeserializeAs<'de, T> for Same {
    fn deserialize_as(value: &mut Value<'de>) -> Result<T, DeserError> {
        T::deserialize(value)
    }
}

impl<'de, T, U: DeserializeAs<'de, T>> DeserializeAs<'de, Spanned<T>> for Spanned<U> {
    fn deserialize_as(value: &mut Value<'de>) -> Result<Spanned<T>, DeserError> {
        let span = value.span;
        let value = U::deserialize_as(value)?;
        Ok(Spanned { value, span })
    }
}

impl<'de, T, U> DeserializeAs<'de, Vec<T>> for Vec<U>
where
    U: DeserializeAs<'de, T>,
{
    fn deserialize_as(value: &mut Value<'de>) -> Result<Vec<T>, DeserError> {
        let array = match value.take() {
            ValueInner::Array(array) => array,
            other => {
                return Err(DeserError::from(expected("array", other, value.span)));
            }
        };

        let mut errors = DeserError { errors: Vec::new() };
        let mut result = Vec::with_capacity(array.len());
        for mut value in array {
            match U::deserialize_as(&mut value) {
                Ok(v) => result.push(v),
                Err(e) => errors.merge(e),
            }
        }

        if errors.errors.is_empty() {
            Ok(result)
        } else {
            Err(errors)
        }
    }
}

pub trait FromKey<'de>: Sized {
    type Err;

    fn from_key(key: toml_span::value::Key<'de>) -> Result<Self, Self::Err>;
}

impl<'de, T: FromStr> FromKey<'de> for T {
    type Err = <T as FromStr>::Err;

    fn from_key(key: toml_span::value::Key<'de>) -> Result<Self, Self::Err> {
        key.name.parse()
    }
}
