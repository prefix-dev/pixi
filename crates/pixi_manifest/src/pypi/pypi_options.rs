use crate::consts;
use indexmap::IndexSet;
use rattler_lock::{FindLinksUrlOrPath, PypiIndexes};
use serde::{Deserialize, Serialize};
use serde_with::serde_as;
use std::{hash::Hash, iter};
use thiserror::Error;
use url::Url;

/// Specific options for a PyPI registries
#[serde_as]
#[derive(Debug, Clone, PartialEq, Serialize, Eq, Deserialize, Default)]
#[serde(rename_all = "kebab-case", deny_unknown_fields)]
pub struct PypiOptions {
    /// The index URL to use as the primary pypi index
    pub index_url: Option<Url>,
    /// Any extra indexes to use, that will be searched after the primary index
    pub extra_index_urls: Option<Vec<Url>>,
    /// Flat indexes also called `--find-links` in pip
    /// These are flat listings of distributions
    pub find_links: Option<Vec<FindLinksUrlOrPath>>,
    /// Disable isolated builds
    pub no_build_isolation: Option<Vec<String>>,
}

/// Clones and deduplicates two iterators of values
fn clone_and_deduplicate<'a, I: Iterator<Item = &'a T>, T: Clone + Eq + Hash + 'a>(
    values: I,
    other: I,
) -> Vec<T> {
    values
        .cloned()
        .chain(other.cloned())
        .collect::<IndexSet<_>>()
        .into_iter()
        .collect::<Vec<_>>()
}

impl PypiOptions {
    pub fn new(
        index: Option<Url>,
        extra_indexes: Option<Vec<Url>>,
        flat_indexes: Option<Vec<FindLinksUrlOrPath>>,
        no_build_isolation: Option<Vec<String>>,
    ) -> Self {
        Self {
            index_url: index,
            extra_index_urls: extra_indexes,
            find_links: flat_indexes,
            no_build_isolation,
        }
    }

    /// Return an iterator over all [`Url`] entries.
    /// In order of:
    /// - `find_links`
    /// - `extra_index_urls`
    /// - `index_url`
    pub fn urls(&self) -> impl Iterator<Item = &'_ Url> + '_ {
        let find_links = self
            .find_links
            .iter()
            .flatten()
            .filter_map(|index| match index {
                FindLinksUrlOrPath::Path(_) => None,
                FindLinksUrlOrPath::Url(url) => Some(url),
            });

        let extra_indexes = self.extra_index_urls.iter().flatten();
        find_links
            .chain(extra_indexes)
            .chain(std::iter::once(self.index_url.as_ref()).flatten())
    }

    /// Merges two `PypiOptions` together, according to the following rules
    /// - There can only be one primary index
    /// - Extra indexes are merged and deduplicated, in the order they are provided
    /// - Flat indexes are merged and deduplicated, in the order they are provided
    pub fn union(&self, other: &PypiOptions) -> Result<PypiOptions, PypiOptionsMergeError> {
        let index = if let Some(other_index) = other.index_url.clone() {
            // Allow only one index
            if let Some(own_index) = self.index_url.clone() {
                return Err(PypiOptionsMergeError::MultiplePrimaryIndexes {
                    first: own_index.to_string(),
                    second: other_index.to_string(),
                });
            } else {
                // Use the other index, because we don't have one
                Some(other_index)
            }
        } else {
            // Use our index, because the other doesn't have one
            self.index_url.clone()
        };

        // Chain together and deduplicate the extra indexes
        let extra_indexes = self
            .extra_index_urls
            .as_ref()
            // Map for value
            .map(|extra_indexes| {
                clone_and_deduplicate(
                    extra_indexes.iter(),
                    other.extra_index_urls.clone().unwrap_or_default().iter(),
                )
            })
            .or_else(|| other.extra_index_urls.clone());

        // Chain together and deduplicate the flat indexes
        let flat_indexes = self
            .find_links
            .as_ref()
            .map(|flat_indexes| {
                clone_and_deduplicate(
                    flat_indexes.iter(),
                    other.find_links.clone().unwrap_or_default().iter(),
                )
            })
            .or_else(|| other.find_links.clone());

        // Merge all the no build isolation packages
        let no_build_isolation = self
            .no_build_isolation
            .as_ref()
            .map(|no_build_isolation| {
                clone_and_deduplicate(
                    no_build_isolation.iter(),
                    other.no_build_isolation.clone().unwrap_or_default().iter(),
                )
            })
            .or_else(|| other.no_build_isolation.clone());

        Ok(PypiOptions {
            index_url: index,
            extra_index_urls: extra_indexes,
            find_links: flat_indexes,
            no_build_isolation,
        })
    }
}

impl From<PypiOptions> for rattler_lock::PypiIndexes {
    fn from(value: PypiOptions) -> Self {
        let primary_index = value
            .index_url
            .unwrap_or(consts::DEFAULT_PYPI_INDEX_URL.clone());
        Self {
            indexes: iter::once(primary_index)
                .chain(value.extra_index_urls.into_iter().flatten())
                .collect(),
            find_links: value.find_links.into_iter().flatten().collect(),
        }
    }
}

impl From<&PypiOptions> for rattler_lock::PypiIndexes {
    fn from(value: &PypiOptions) -> Self {
        PypiIndexes::from(value.clone())
    }
}

#[derive(Error, Debug)]
pub enum PypiOptionsMergeError {
    #[error(
        "multiple primary pypi indexes are not supported, found both {first} and {second} across multiple pypi options"
    )]
    MultiplePrimaryIndexes { first: String, second: String },
}

#[cfg(test)]
mod tests {
    use super::PypiOptions;
    use rattler_lock::FindLinksUrlOrPath;
    use url::Url;

    #[test]
    fn test_deserialize_pypi_options() {
        let toml_str = r#"
                 index-url = "https://example.com/pypi"
                 extra-index-urls = ["https://example.com/extra"]
                 no-build-isolation = ["pkg1", "pkg2"]

                 [[find-links]]
                 path = "/path/to/flat/index"

                 [[find-links]]
                 url = "https://flat.index"
             "#;
        let deserialized_options: PypiOptions = toml_edit::de::from_str(toml_str).unwrap();
        assert_eq!(
            deserialized_options,
            PypiOptions {
                index_url: Some(Url::parse("https://example.com/pypi").unwrap()),
                extra_index_urls: Some(vec![Url::parse("https://example.com/extra").unwrap()]),
                find_links: Some(vec![
                    FindLinksUrlOrPath::Path("/path/to/flat/index".into()),
                    FindLinksUrlOrPath::Url(Url::parse("https://flat.index").unwrap())
                ]),
                no_build_isolation: Some(vec!["pkg1".to_string(), "pkg2".to_string()])
            },
        );
    }

    #[test]
    fn test_merge_pypi_options() {
        // Create the first set of options
        let opts = PypiOptions {
            index_url: Some(Url::parse("https://example.com/pypi").unwrap()),
            extra_index_urls: Some(vec![Url::parse("https://example.com/extra").unwrap()]),
            find_links: Some(vec![
                FindLinksUrlOrPath::Path("/path/to/flat/index".into()),
                FindLinksUrlOrPath::Url(Url::parse("https://flat.index").unwrap()),
            ]),
            no_build_isolation: Some(vec!["foo".to_string(), "bar".to_string()]),
        };

        // Create the second set of options
        let opts2 = PypiOptions {
            index_url: None,
            extra_index_urls: Some(vec![Url::parse("https://example.com/extra2").unwrap()]),
            find_links: Some(vec![
                FindLinksUrlOrPath::Path("/path/to/flat/index2".into()),
                FindLinksUrlOrPath::Url(Url::parse("https://flat.index2").unwrap()),
            ]),
            no_build_isolation: Some(vec!["foo".to_string()]),
        };

        // Merge the two options
        // This should succeed and values should be merged
        let merged_opts = opts.union(&opts2).unwrap();
        insta::assert_yaml_snapshot!(merged_opts);
    }

    #[test]
    fn test_error_on_multiple_primary_indexes() {
        // Create the first set of options
        let opts = PypiOptions {
            index_url: Some(Url::parse("https://example.com/pypi").unwrap()),
            extra_index_urls: None,
            find_links: None,
            no_build_isolation: None,
        };

        // Create the second set of options
        let opts2 = PypiOptions {
            index_url: Some(Url::parse("https://example.com/pypi2").unwrap()),
            extra_index_urls: None,
            find_links: None,
            no_build_isolation: None,
        };

        // Merge the two options
        // This should error because there are two primary indexes
        let merged_opts = opts.union(&opts2);
        insta::assert_snapshot!(merged_opts.err().unwrap());
    }
}
