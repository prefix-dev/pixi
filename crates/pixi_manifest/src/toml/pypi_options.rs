use std::{collections::HashSet, path::PathBuf};

use pixi_toml::{TomlEnum, TomlFromStr, TomlWith};
use toml_span::{
    de_helpers::{expected, TableHelper},
    value::ValueInner,
    DeserError, ErrorKind, Value,
};
use url::Url;

use crate::pypi::pypi_options::{FindLinksUrlOrPath, NoBuild, PypiOptions};

impl<'de> toml_span::Deserialize<'de> for NoBuild {
    fn deserialize(value: &mut Value<'de>) -> Result<Self, DeserError> {
        // It can be either `true` or `false` or an array of strings
        if value.as_bool().is_some() {
            if bool::deserialize(value)? {
                return Ok(NoBuild::All);
            } else {
                return Ok(NoBuild::None);
            }
        }
        // We assume it's an array of strings
        if value.as_array().is_some() {
            match value.take() {
                ValueInner::Array(array) => {
                    let mut packages = HashSet::with_capacity(array.len());
                    for mut value in array {
                        packages.insert(value.take_string(None)?.into_owned());
                    }
                    Ok(NoBuild::Packages(packages))
                }
                _ => Err(expected(
                    "an array of packages e.g. [\"foo\", \"bar\"]",
                    value.take(),
                    value.span,
                )
                .into()),
            }
        } else {
            Err(expected(
                r#"either "all", "none" or an array of packages e.g. ["foo", "bar"] "#,
                value.take(),
                value.span,
            )
            .into())
        }
    }
}

impl<'de> toml_span::Deserialize<'de> for PypiOptions {
    fn deserialize(value: &mut Value<'de>) -> Result<Self, DeserError> {
        let mut th = TableHelper::new(value)?;

        let index_url = th
            .optional::<TomlFromStr<_>>("index-url")
            .map(TomlFromStr::into_inner);
        let extra_index_urls = th
            .optional::<TomlWith<_, Vec<TomlFromStr<_>>>>("extra-index-urls")
            .map(|x| x.into_inner());
        let find_links = th.optional("find-links");
        let no_build_isolation = th.optional("no-build-isolation");
        let index_strategy = th
            .optional::<TomlEnum<_>>("index-strategy")
            .map(TomlEnum::into_inner);

        let no_build = th.optional::<NoBuild>("no-build");

        th.finalize(None)?;

        Ok(Self {
            index_url,
            extra_index_urls,
            find_links,
            no_build_isolation,
            index_strategy,
            no_build,
        })
    }
}

impl<'de> toml_span::Deserialize<'de> for FindLinksUrlOrPath {
    fn deserialize(value: &mut Value<'de>) -> Result<Self, DeserError> {
        let mut table = match value.take() {
            ValueInner::Table(table) => table,
            other => return Err(expected("a table", other, value.span).into()),
        };

        let mut errors = DeserError { errors: vec![] };

        let path = table.remove("path");
        let url = table.remove("url");
        if path.is_some() && url.is_some() {
            errors.errors.push(toml_span::Error {
                kind: ErrorKind::Custom("cannot define both 'url' and 'path'".into()),
                span: value.span,
                line_info: None,
            });
        }

        let path = if let Some(mut path) = path {
            match path
                .take_string(None)
                .map(|str| PathBuf::from(str.into_owned()))
            {
                Err(error) => {
                    errors.errors.push(error);
                    None
                }
                Ok(path) => Some(path),
            }
        } else {
            None
        };

        let url = if let Some(mut url) = url {
            match url.take_string(None).map(|str| Url::parse(&str)) {
                Err(error) => {
                    errors.errors.push(error);
                    None
                }
                Ok(Err(error)) => {
                    errors.errors.push(toml_span::Error {
                        kind: ErrorKind::Custom(error.to_string().into()),
                        span: url.span,
                        line_info: None,
                    });
                    None
                }
                Ok(Ok(url)) => Some(url),
            }
        } else {
            None
        };

        if !errors.errors.is_empty() {
            return Err(errors);
        }

        if let Some(url) = url {
            Ok(Self::Url(url))
        } else if let Some(path) = path {
            Ok(Self::Path(path))
        } else {
            Err(DeserError::from(toml_span::Error {
                kind: ErrorKind::Custom("either 'url' or 'path' must be defined".into()),
                span: value.span,
                line_info: None,
            }))
        }
    }
}

#[cfg(test)]
mod test {
    use insta::{assert_debug_snapshot, assert_snapshot};

    use super::*;
    use crate::{toml::FromTomlStr, utils::test_utils::format_parse_error};

    #[test]
    fn test_empty() {
        let input = "";
        let options = PypiOptions::from_toml_str(input).unwrap();
        assert_eq!(options, PypiOptions::default());
    }

    #[test]
    fn test_deserialize_pypi_options() {
        let toml_str = r#"
                 index-url = "https://example.com/pypi"
                 extra-index-urls = ["https://example.com/extra"]
                 no-build-isolation = ["pkg1", "pkg2"]

                 [[find-links]]
                 path = "/path/to/flat/index"

                 [[find-links]]
                 url = "https://flat.index"
             "#;
        let deserialized_options: PypiOptions = PypiOptions::from_toml_str(toml_str).unwrap();
        assert_eq!(
            deserialized_options,
            PypiOptions {
                index_url: Some(Url::parse("https://example.com/pypi").unwrap()),
                extra_index_urls: Some(vec![Url::parse("https://example.com/extra").unwrap()]),
                find_links: Some(vec![
                    FindLinksUrlOrPath::Path("/path/to/flat/index".into()),
                    FindLinksUrlOrPath::Url(Url::parse("https://flat.index").unwrap())
                ]),
                no_build_isolation: Some(vec!["pkg1".to_string(), "pkg2".to_string()]),
                index_strategy: None,
                no_build: Default::default(),
            },
        );
    }

    #[test]
    fn test_full() {
        let input = r#"
        index-url = "https://pypi.org/simple"
        extra-index-urls = ["https://pypi.org/simple", "file:///path/to/simple"]
        find-links = [
            { path = "../" },
            { url = "https://google.com" }
        ]
        no-build-isolation = ["sigma"]
        index-strategy = "first-index"
        no-build = true
        "#;
        let options = PypiOptions::from_toml_str(input).unwrap();
        assert_debug_snapshot!(options);
    }

    #[test]
    fn test_no_build_packages() {
        let input = r#"
        no-build = ["package1"]
        "#;
        let options = PypiOptions::from_toml_str(input).unwrap();
        assert_debug_snapshot!(options);
    }

    #[test]
    fn test_invalid_strategy_missing_dash() {
        let input = r#"
        index-strategy = "firstindex"
        "#;
        assert_snapshot!(
            format_parse_error(input, PypiOptions::from_toml_str(input).unwrap_err()),
            @r###"
         × Expected one of 'first-index', 'unsafe-first-match', 'unsafe-best-match'
          ╭─[pixi.toml:2:27]
        1 │
        2 │         index-strategy = "firstindex"
          ·                           ──────────
        3 │
          ╰────
         help: Did you mean 'first-index'?
        "###
        )
    }

    #[test]
    fn test_invalid_strategy_upper_case() {
        let input = r#"
        index-strategy = "UnsafeFirstMatch"
        "#;
        assert_snapshot!(
            format_parse_error(input, PypiOptions::from_toml_str(input).unwrap_err()),
            @r###"
         × Expected one of 'first-index', 'unsafe-first-match', 'unsafe-best-match'
          ╭─[pixi.toml:2:27]
        1 │
        2 │         index-strategy = "UnsafeFirstMatch"
          ·                           ────────────────
        3 │
          ╰────
         help: Did you mean 'unsafe-first-match'?
        "###
        )
    }

    #[test]
    fn test_invalid_strategy_far() {
        let input = r#"
        index-strategy = "foobar"
        "#;
        assert_snapshot!(
            format_parse_error(input, PypiOptions::from_toml_str(input).unwrap_err()),
            @r###"
         × Expected one of 'first-index', 'unsafe-first-match', 'unsafe-best-match'
          ╭─[pixi.toml:2:27]
        1 │
        2 │         index-strategy = "foobar"
          ·                           ──────
        3 │
          ╰────
        "###
        )
    }

    #[test]
    fn test_missing_url_or_path() {
        let input = "find-links = [{}]";
        assert_snapshot!(
            format_parse_error(input, PypiOptions::from_toml_str(input).unwrap_err()),
            @r###"
         × either 'url' or 'path' must be defined
          ╭─[pixi.toml:1:15]
        1 │ find-links = [{}]
          ·               ──
          ╰────
        "###
        )
    }

    #[test]
    fn test_both_url_or_path() {
        let input = r#"find-links = [{url = "", path = ""}]"#;
        assert_snapshot!(
            format_parse_error(input, PypiOptions::from_toml_str(input).unwrap_err()),
            @r###"
         × cannot define both 'url' and 'path'
          ╭─[pixi.toml:1:15]
        1 │ find-links = [{url = "", path = ""}]
          ·               ─────────────────────
          ╰────
        "###
        )
    }

    #[test]
    fn test_wrong_build_option_type() {
        let input = r#"no-build = 3"#;
        assert_snapshot!(format_parse_error(
            input,
            PypiOptions::from_toml_str(input).unwrap_err()
        ), @r###"
         × expected either "all", "none" or an array of packages e.g. ["foo", "bar"] , found integer
          ╭─[pixi.toml:1:12]
        1 │ no-build = 3
          ·            ─
          ╰────
        "###)
    }
}
