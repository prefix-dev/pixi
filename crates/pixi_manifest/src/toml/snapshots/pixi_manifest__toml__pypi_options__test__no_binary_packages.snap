---
source: crates/pixi_manifest/src/toml/pypi_options.rs
assertion_line: 317
expression: options
---
PypiOptions {
    index_url: None,
    extra_index_urls: None,
    find_links: None,
    no_build_isolation: Packages(
        {},
    ),
    index_strategy: None,
    no_build: None,
    no_binary: Some(
        Packages(
            {
                PackageName(
                    "package1",
                ),
            },
        ),
    ),
}
